src/ceph-volume/ceph_volume/devices/lvm/batch.py

```diff
@@ -316,79 +387,216 @@ def main(self):
-                # only fail if non-interactive, this iteration concerns
-                # non-data devices, there are usable data devices (or not all
-                # data devices were filtered) and non-data devices were filtered
-                # so in short this branch is not taken if all data devices are
-                # filtered
-                if self.args.yes and dev_list and self.usable and devs != usable:
-                    err = '{} devices were filtered in non-interactive mode, bailing out'
-                    if self.args.format == "json" and self.args.report:
-                        # if a json report is requested, report unchanged so idempotency checks
-                        # in ceph-ansible will work
-                        print(json.dumps({"changed": False, "osds": [], "vgs": []}))
-                        raise SystemExit(0)
-                    raise RuntimeError(err.format(len(devs) - len(usable)))
-
-
-    def _ensure_disjoint_device_lists(self):
-        # check that all device lists are disjoint with each other
-        if not(set(self.usable).isdisjoint(set(self.db_usable)) and
-               set(self.usable).isdisjoint(set(self.wal_usable)) and
-               set(self.usable).isdisjoint(set(self.journal_usable)) and
-               set(self.db_usable).isdisjoint(set(self.wal_usable))):
-            raise Exception('Device lists are not disjoint')
+        self._execute(plan)
+
+    def _execute(self, plan):
+        defaults = common.get_default_args()
+        global_args = [
+            'bluestore',
+            'filestore',
+            'dmcrypt',
+            'crush_device_class',
+            'no_systemd',
+        ]
+        defaults.update({arg: getattr(self.args, arg) for arg in global_args})
+        for osd in plan:
+            args = osd.get_args(defaults)
+            if self.args.prepare:
+                p = Prepare([])
+                p.prepare(argparse.Namespace(**args))
+            else:
+                c = Create([])
+                c.create(argparse.Namespace(**args))
+
+
+    def get_plan(self, args):
+        if args.bluestore:
+            plan = self.get_deployment_layout(args, args.devices, args.db_devices,
+                                              args.wal_devices)
+        elif args.filestore:
+            plan = self.get_deployment_layout(args, args.devices, args.journal_devices)
+        return plan
+
+    def get_deployment_layout(self, args, devices, fast_devices=[],
+                              very_fast_devices=[]):
+        '''
+        The methods here are mostly just organization, error reporting and
+        setting up of (default) args. The hravy lifting code for the deployment
+        layout can be found in the static get_*_osds and get_*:fast_allocs
+        functions.
+        '''
+        plan = []
+        phys_devs = [d for d in devices if d.is_device]
+        lvm_devs = [d.lvs[0] for d in list(set(devices) -
```

Comment:
Is there only ever one device under `lvs`. According to

 https://github.com/ceph/ceph/blob/master/src/ceph-volume/ceph_volume/api/lvm.py#L1099-L1106 

it can have multiple.

Reply:
We get `util/device::Device()` objects here. When these are created from an `vg/lv` argument `lvs` will be a single item list.
See https://github.com/ceph/ceph/blob/master/src/ceph-volume/ceph_volume/util/device.py#L153-L161

This class is also in need of serious overhaul, but one giant PR at a time.